File: source/src/utils/googleDocsChecklist.js
/**
 * DOCX→GDoc 変換後、段落先頭の ☐ / ☑ マーカーを削除し、
 * Google Docs API のネイティブチェックリスト（BULLET_CHECKBOX）に置き換える。
 *
 * ☑ 行は API が箇条書きの「チェック済み」状態を設定できないため、ラベルに取り消し線を付けて完了を表す。
 *
 * @see https://developers.google.com/docs/api/how-tos/lists
 */

const UNCHECKED = '\u2610'; // BALLOT BOX（docx 側の未チェックマーカー）
const CHECKED = '\u2611'; // BALLOT BOX WITH CHECK（docx 側のチェック済みマーカー）

/** GCP で Google Docs API が無効なとき、全ファイルで同じ 403 が繰り返されるのを防ぐ */
let docsApiServiceDisabled = false;

function isDocsApiServiceDisabled(status, bodyText) {
  if (status !== 403 || !bodyText) return false;
  try {
    const data = JSON.parse(bodyText);
    const err = data.error || {};
    const details = err.details || [];
    for (const d of details) {
      if (d.reason === 'SERVICE_DISABLED') return true;
    }
    const msg = (err.message || '').toLowerCase();
    if (msg.includes('docs.googleapis.com') && msg.includes('disabled')) return true;
  } catch {
    /* ignore */
  }
  return false;
}

function walkStructuralElements(elements, out) {
  if (!elements) return;
  for (const el of elements) {
    if (el.paragraph) out.push(el);
    if (el.table) {
      for (const row of el.table.tableRows || []) {
        for (const cell of row.tableCells || []) {
          walkStructuralElements(cell.content, out);
        }
      }
    }
  }
}

function paragraphPlainText(paragraph) {
  return (paragraph.elements || []).map((e) => e.textRun?.content || '').join('');
}

/** Google Docs のインデックスは UTF-16 コードユニット。連結テキスト（textRun のみ）のオフセットを絶対 startIndex に変換 */
function mapUtf16OffsetToDocumentIndex(paragraph, utf16Offset) {
  let u = 0;
  for (const el of paragraph.elements || []) {
    const tr = el.textRun;
    if (!tr?.content) continue;
    const content = tr.content;
    if (utf16Offset < u + content.length) {
      return el.startIndex + (utf16Offset - u);
    }
    u += content.length;
  }
  return null;
}

/** ☐ の直後に続く区切り（通常スペース・NBSP・改行・タブ・薄スペース・和字間隔など） */
function isSpaceAfterCheckbox(ch) {
  return /[ \n\r\t\u00a0\u2009\u3000]/.test(ch);
}

/**
 * @param {string} documentId
 * @param {string} accessToken
 * @returns {Promise<boolean>} 成功したら true
 */
export async function convertTaskMarkersToNativeChecklists(documentId, accessToken) {
  if (!documentId || !accessToken) return false;
  if (docsApiServiceDisabled) return false;

  const getRes = await fetch(`https://docs.googleapis.com/v1/documents/${documentId}`, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  const getBody = await getRes.text();
  if (!getRes.ok) {
    if (isDocsApiServiceDisabled(getRes.status, getBody)) {
      if (!docsApiServiceDisabled) {
        docsApiServiceDisabled = true;
        console.warn(
          '[GDoc] Google Docs API が GCP で無効です（チェックリストのネイティブ化をスキップ）。' +
            ' 有効化: https://console.developers.google.com/apis/library/docs.googleapis.com'
        );
      }
      return false;
    }
    console.warn('[GDoc] documents.get failed', getRes.status, getBody);
    return false;
  }

  const doc = JSON.parse(getBody);
  const structural = [];
  walkStructuralElements(doc.body?.content, structural);

  /** @type {{ firstIdx: number, seEnd: number, deleteLen: number }[]} */
  const candidates = [];
  /** 変換前の段落インデックス（structural 内）— ☑ 行は後段で取り消し線 */
  const checkedParagraphIndices = [];

  for (let j = 0; j < structural.length; j++) {
    const se = structural[j];
    const p = se.paragraph;
    if (!p?.elements?.length) continue;
    const text = paragraphPlainText(p);
    // インデントが別 TextRun（タブ等）のとき elements[0] は ☐ を含まない → 連結テキスト上で位置を解決
    const leadMatch = text.match(/^[\t \u00a0\u2009\u3000]*/);
    const leadLen = leadMatch ? leadMatch[0].length : 0;
    const afterLead = text.slice(leadLen);
    let marker = null;
    if (afterLead.startsWith(UNCHECKED)) marker = UNCHECKED;
    else if (afterLead.startsWith(CHECKED)) marker = CHECKED;
    else continue;
    const afterBox = afterLead.slice(marker.length);
    if (afterBox.length === 0) continue;
    if (!isSpaceAfterCheckbox(afterBox[0])) continue;

    const deleteLen = 2; // マーカー + 直後1文字（BMP 想定）
    const deleteStart = mapUtf16OffsetToDocumentIndex(p, leadLen);
    if (deleteStart == null || se.endIndex == null) continue;
    if (se.endIndex - deleteStart <= deleteLen) continue;

    candidates.push({ firstIdx: deleteStart, seEnd: se.endIndex, deleteLen });
    if (marker === CHECKED) checkedParagraphIndices.push(j);
  }

  if (candidates.length === 0) return true;

  candidates.sort((a, b) => b.firstIdx - a.firstIdx);

  const requests = [];
  for (const c of candidates) {
    requests.push({
      deleteContentRange: {
        range: { startIndex: c.firstIdx, endIndex: c.firstIdx + c.deleteLen },
      },
    });
    requests.push({
      createParagraphBullets: {
        range: { startIndex: c.firstIdx, endIndex: c.seEnd - c.deleteLen },
        bulletPreset: 'BULLET_CHECKBOX',
      },
    });
  }

  const batchRes = await fetch(
    `https://docs.googleapis.com/v1/documents/${documentId}:batchUpdate`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ requests }),
    }
  );

  const batchBody = await batchRes.text();
  if (!batchRes.ok) {
    if (isDocsApiServiceDisabled(batchRes.status, batchBody)) {
      if (!docsApiServiceDisabled) {
        docsApiServiceDisabled = true;
        console.warn(
          '[GDoc] Google Docs API が GCP で無効です（チェックリストのネイティブ化をスキップ）。' +
            ' 有効化: https://console.developers.google.com/apis/library/docs.googleapis.com'
        );
      }
      return false;
    }
    console.warn('[GDoc] batchUpdate failed', batchRes.status, batchBody);
    return false;
  }

  // ☑ 行: ネイティブの「チェック済み」状態は API 未対応のため、ラベルに取り消し線を付ける
  if (checkedParagraphIndices.length === 0) return true;

  const get2 = await fetch(`https://docs.googleapis.com/v1/documents/${documentId}`, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  const get2Body = await get2.text();
  if (!get2.ok) {
    console.warn('[GDoc] documents.get (after checklist) failed', get2.status, get2Body);
    return true;
  }
  const doc2 = JSON.parse(get2Body);
  const structural2 = [];
  walkStructuralElements(doc2.body?.content, structural2);
  if (structural2.length !== structural.length) {
    console.warn('[GDoc] paragraph count changed after checklist batch; skip strikethrough');
    return true;
  }

  const strikeRequests = [];
  for (const idx of checkedParagraphIndices) {
    const se = structural2[idx];
    const p = se?.paragraph;
    const els = p?.elements;
    if (!els?.length) continue;
    const startIdx = els[0].startIndex;
    const endIdx = els[els.length - 1].endIndex;
    if (startIdx == null || endIdx == null || endIdx <= startIdx) continue;
    strikeRequests.push({
      updateTextStyle: {
        range: { startIndex: startIdx, endIndex: endIdx },
        textStyle: { strikethrough: true },
        fields: 'strikethrough',
      },
    });
  }
  if (strikeRequests.length === 0) return true;

  const strikeRes = await fetch(
    `https://docs.googleapis.com/v1/documents/${documentId}:batchUpdate`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ requests: strikeRequests }),
    }
  );
  if (!strikeRes.ok) {
    const err = await strikeRes.text();
    console.warn('[GDoc] batchUpdate (strikethrough for checked tasks) failed', strikeRes.status, err);
  }
  return true;
}

export { UNCHECKED, CHECKED };

File: desktop/app/google_docs_checklist.py
"""DOCX→GDoc 後、☐/☑ マーカーを Google Docs API のネイティブチェックリスト（BULLET_CHECKBOX）にする。

☑ 行は API が箇条書きの「チェック済み」状態を設定できないため、ラベルに取り消し線を付ける。
source/src/utils/googleDocsChecklist.js と同じ方針。
"""

from __future__ import annotations

import json
import logging
import re
from typing import Any

import httpx

from .google_oauth_refresh import google_request_with_token_refresh

logger = logging.getLogger(__name__)

UNCHECKED = "\u2610"
CHECKED = "\u2611"

# Google Docs API が GCP で無効（SERVICE_DISABLED）のとき、全ファイルで同じ 403 が繰り返されるのを防ぐ
_docs_api_unavailable: bool = False


def _is_docs_api_service_disabled(status_code: int, body: str) -> bool:
    if status_code != 403:
        return False
    try:
        data = json.loads(body)
        err = data.get("error") or {}
        for d in err.get("details") or []:
            if d.get("reason") == "SERVICE_DISABLED":
                return True
        msg = (err.get("message") or "").lower()
        if "docs.googleapis.com" in msg and "disabled" in msg:
            return True
    except (json.JSONDecodeError, TypeError):
        pass
    return False


# 先頭インデント・☐/☑ 直後の区切り（GDoc が NBSP / 全角スペースにすることがある）
_TASK_MARKER_LINE = re.compile(
    r"^([\t \u00a0\u2009\u3000]*)([\u2610\u2611])([ \u00a0\n\r\t\u2009\u3000])"
)


def _utf16_len(s: str) -> int:
    return sum(2 if ord(c) > 0xFFFF else 1 for c in s)


def _map_utf16_offset_to_document_index(
    elems: list[dict[str, Any]], utf16_off: int
) -> int | None:
    u = 0
    for e in elems:
        tr = e.get("textRun") or {}
        content = tr.get("content") or ""
        if not content:
            continue
        n = _utf16_len(content)
        if utf16_off < u + n:
            si = e.get("startIndex")
            if si is None:
                return None
            return si + (utf16_off - u)
        u += n
    return None


def _flatten_paragraph_structural_elements(
    elements: list | None, out: list[dict[str, Any]]
) -> None:
    if not elements:
        return
    for el in elements:
        if "paragraph" in el:
            out.append(el)
        elif "table" in el:
            for row in (el.get("table") or {}).get("tableRows") or []:
                for cell in row.get("tableCells") or []:
                    _flatten_paragraph_structural_elements(cell.get("content"), out)


async def convert_task_markers_to_native_checklists(
    client: httpx.AsyncClient,
    token_ref: list[str],
    refresh: str | None,
    document_id: str,
) -> bool:
    global _docs_api_unavailable

    if not document_id or not token_ref or not token_ref[0]:
        return False
    if _docs_api_unavailable:
        return False

    r = await google_request_with_token_refresh(
        client,
        token_ref,
        refresh,
        lambda tok: client.get(
            f"https://docs.googleapis.com/v1/documents/{document_id}",
            headers={"Authorization": f"Bearer {tok}"},
            timeout=120.0,
        ),
    )
    if not r.is_success:
        body = r.text or ""
        if _is_docs_api_service_disabled(r.status_code, body):
            if not _docs_api_unavailable:
                _docs_api_unavailable = True
                logger.warning(
                    "Google Docs API が GCP プロジェクトで無効のため、チェックリストのネイティブ化をスキップします。"
                    " コンソールで「Google Docs API」を有効化してください: "
                    "https://console.developers.google.com/apis/library/docs.googleapis.com"
                )
            return False
        logger.warning(
            "documents.get failed status=%s body=%s",
            r.status_code,
            body[:2000],
        )
        return False

    doc: dict[str, Any] = r.json()
    structural_flat: list[dict[str, Any]] = []
    _flatten_paragraph_structural_elements(doc.get("body", {}).get("content"), structural_flat)

    candidates: list[dict[str, int]] = []
    checked_paragraph_indices: list[int] = []

    for j, el in enumerate(structural_flat):
        p = el.get("paragraph") or {}
        elems = p.get("elements") or []
        if not elems:
            continue
        text = "".join(
            (e.get("textRun") or {}).get("content") or "" for e in elems
        )
        m = _TASK_MARKER_LINE.match(text)
        if not m:
            continue
        marker = m.group(2)
        marker_start_cp = m.start(2)
        utf16_off = _utf16_len(text[:marker_start_cp])
        delete_len = 2
        first_idx = _map_utf16_offset_to_document_index(elems, utf16_off)
        se_end = el.get("endIndex")
        if first_idx is None or se_end is None:
            continue
        if se_end - first_idx <= delete_len:
            continue
        candidates.append(
            {"firstIdx": first_idx, "seEnd": se_end, "deleteLen": delete_len}
        )
        if marker == CHECKED:
            checked_paragraph_indices.append(j)

    if not candidates:
        return True

    candidates.sort(key=lambda c: c["firstIdx"], reverse=True)
    requests: list[dict[str, Any]] = []
    for c in candidates:
        fi, se_end, dl = c["firstIdx"], c["seEnd"], c["deleteLen"]
        requests.append(
            {
                "deleteContentRange": {
                    "range": {"startIndex": fi, "endIndex": fi + dl},
                }
            }
        )
        requests.append(
            {
                "createParagraphBullets": {
                    "range": {"startIndex": fi, "endIndex": se_end - dl},
                    "bulletPreset": "BULLET_CHECKBOX",
                }
            }
        )

    br = await google_request_with_token_refresh(
        client,
        token_ref,
        refresh,
        lambda tok: client.post(
            f"https://docs.googleapis.com/v1/documents/{document_id}:batchUpdate",
            headers={
                "Authorization": f"Bearer {tok}",
                "Content-Type": "application/json",
            },
            json={"requests": requests},
            timeout=120.0,
        ),
    )
    if not br.is_success:
        berr = br.text or ""
        if _is_docs_api_service_disabled(br.status_code, berr):
            if not _docs_api_unavailable:
                _docs_api_unavailable = True
                logger.warning(
                    "Google Docs API が GCP プロジェクトで無効のため、チェックリストのネイティブ化をスキップします。"
                    " コンソールで「Google Docs API」を有効化してください: "
                    "https://console.developers.google.com/apis/library/docs.googleapis.com"
                )
            return False
        logger.warning(
            "documents.batchUpdate failed status=%s body=%s",
            br.status_code,
            berr[:2000],
        )
        return False

    if not checked_paragraph_indices:
        return True

    r2 = await google_request_with_token_refresh(
        client,
        token_ref,
        refresh,
        lambda tok: client.get(
            f"https://docs.googleapis.com/v1/documents/{document_id}",
            headers={"Authorization": f"Bearer {tok}"},
            timeout=120.0,
        ),
    )
    if not r2.is_success:
        logger.warning(
            "documents.get (after checklist) failed status=%s body=%s",
            r2.status_code,
            (r2.text or "")[:2000],
        )
        return True

    doc2: dict[str, Any] = r2.json()
    structural_flat2: list[dict[str, Any]] = []
    _flatten_paragraph_structural_elements(
        doc2.get("body", {}).get("content"), structural_flat2
    )
    if len(structural_flat2) != len(structural_flat):
        logger.warning(
            "paragraph count changed after checklist batch; skip strikethrough for checked tasks"
        )
        return True

    strike_requests: list[dict[str, Any]] = []
    for idx in checked_paragraph_indices:
        el = structural_flat2[idx]
        p = el.get("paragraph") or {}
        elems = p.get("elements") or []
        if not elems:
            continue
        start_idx = elems[0].get("startIndex")
        end_idx = elems[-1].get("endIndex")
        if start_idx is None or end_idx is None or end_idx <= start_idx:
            continue
        strike_requests.append(
            {
                "updateTextStyle": {
                    "range": {"startIndex": start_idx, "endIndex": end_idx},
                    "textStyle": {"strikethrough": True},
                    "fields": "strikethrough",
                }
            }
        )
    if not strike_requests:
        return True

    sr = await google_request_with_token_refresh(
        client,
        token_ref,
        refresh,
        lambda tok: client.post(
            f"https://docs.googleapis.com/v1/documents/{document_id}:batchUpdate",
            headers={
                "Authorization": f"Bearer {tok}",
                "Content-Type": "application/json",
            },
            json={"requests": strike_requests},
            timeout=120.0,
        ),
    )
    if not sr.is_success:
        logger.warning(
            "documents.batchUpdate (strikethrough) failed status=%s body=%s",
            sr.status_code,
            (sr.text or "")[:2000],
        )
    return True
